File: watch/src/main/java/com/example/wearbabymonitor/NoiseMonitorService.java
package com.example.wearbabymonitor;

import android.app.Notification;
import android.app.NotificationChannel;
import android.app.NotificationManager;
import android.app.PendingIntent;
import android.app.Service;
import android.content.Intent;
import android.content.IntentFilter;
import android.media.AudioFormat;
import android.media.AudioRecord;
import android.media.MediaRecorder;
import android.os.BatteryManager;
import android.os.IBinder;
import android.os.SystemClock;

import com.google.android.gms.wearable.Wearable;

import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Locale;
import java.util.UUID;

public final class NoiseMonitorService extends Service {
    private static final String CHANNEL_ID = "baby_monitor_running_v3";
    private static final int NOTIFICATION_ID = 1;
    private static final double DEFAULT_THRESHOLD = 0.075;
    private static final double MIN_THRESHOLD = 0.018;
    private static final double BASELINE_MULTIPLIER = 3.0;
    private static final long CALIBRATION_MS = 8000L;
    private static final int REQUIRED_LOUD_WINDOWS = 3;
    private static final long ALERT_COOLDOWN_MS = 10000L;
    private static final long STATUS_INTERVAL_MS = 10000L;
    private static final int MAX_SEND_ATTEMPTS = 6;
    private static final long RETRY_INTERVAL_MS = 1500L;
    private static final int LOW_BATTERY_PERCENT = 20;
    private static final int LOW_BATTERY_RESET_PERCENT = 25;

    private volatile boolean running;
    private AudioRecord recorder;
    private Thread worker;
    private long lastAlertAt;
    private double calibratedThreshold = DEFAULT_THRESHOLD;
    private long lastStatusAt;
    private boolean warnedDisconnected;
    private boolean warnedLowBattery;

    @Override
    public void onCreate() {
        super.onCreate();
        createSilentChannel();
    }

    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {
        startForeground(NOTIFICATION_ID, notification("Calibrating room noise…"));
        if (!running) startRecordingLoop();
        return START_NOT_STICKY;
    }

    @Override
    public void onDestroy() {
        boolean wasRunning = running;
        running = false;
        stopRecorderSafely();
        if (recorder != null) recorder.release();
        recorder = null;
        if (worker != null) worker.interrupt();
        worker = null;
        if (wasRunning) sendStatus(false, "stopped");
        stopForeground(STOP_FOREGROUND_REMOVE);
        super.onDestroy();
    }

    @Override
    public IBinder onBind(Intent intent) {
        return null;
    }

    private void startRecordingLoop() {
        final int sampleRate = 16000;
        int minBuffer = AudioRecord.getMinBufferSize(
                sampleRate,
                AudioFormat.CHANNEL_IN_MONO,
                AudioFormat.ENCODING_PCM_16BIT
        );
        if (minBuffer <= 0) {
            failMonitoring("Microphone unavailable", "microphone_unavailable");
            return;
        }

        final int bufferSize = Math.max(minBuffer, 2048);
        AudioRecord audioRecord;
        try {
            audioRecord = new AudioRecord(
                    MediaRecorder.AudioSource.MIC,
                    sampleRate,
                    AudioFormat.CHANNEL_IN_MONO,
                    AudioFormat.ENCODING_PCM_16BIT,
                    bufferSize * 2
            );
        } catch (SecurityException denied) {
            failMonitoring("Microphone permission lost", "permission_lost");
            return;
        }

        if (audioRecord.getState() != AudioRecord.STATE_INITIALIZED) {
            audioRecord.release();
            failMonitoring("Microphone unavailable", "microphone_unavailable");
            return;
        }

        recorder = audioRecord;
        running = true;
        worker = new Thread(() -> recordLoop(audioRecord, bufferSize), "baby-noise-monitor");
        worker.start();
    }

    private void recordLoop(AudioRecord audioRecord, int bufferSize) {
        short[] buffer = new short[bufferSize];
        List<Double> calibrationSamples = new ArrayList<>();
        long calibrationEnd = SystemClock.elapsedRealtime() + CALIBRATION_MS;
        int loudWindows = 0;

        try {
            audioRecord.startRecording();
            sendStatus(true, "calibrating");
            lastStatusAt = SystemClock.elapsedRealtime();

            while (running && !Thread.currentThread().isInterrupted()) {
                int count = audioRecord.read(buffer, 0, buffer.length);
                if (count <= 0) continue;

                double rms = calculateNormalizedRms(buffer, count);
                long nowElapsed = SystemClock.elapsedRealtime();

                if (nowElapsed < calibrationEnd) {
                    calibrationSamples.add(rms);
                    continue;
                }

                if (!calibrationSamples.isEmpty()) {
                    Collections.sort(calibrationSamples);
                    int index = Math.min(
                            calibrationSamples.size() - 1,
                            (int) (calibrationSamples.size() * 0.8)
                    );
                    double baseline = calibrationSamples.get(index);
                    calibratedThreshold = Math.max(MIN_THRESHOLD, baseline * BASELINE_MULTIPLIER);
                    calibrationSamples.clear();
                    updateNotification("Monitoring • threshold "
                            + String.format(Locale.US, "%.3f", calibratedThreshold));
                    sendStatus(true, "monitoring");
                }

                loudWindows = rms >= calibratedThreshold ? loudWindows + 1 : 0;
                if (loudWindows >= REQUIRED_LOUD_WINDOWS) {
                    long now = System.currentTimeMillis();
                    if (now - lastAlertAt >= ALERT_COOLDOWN_MS) {
                        lastAlertAt = now;
                        deliverAlertWithRetry(rms);
                    }
                    loudWindows = 0;
                }

                if (nowElapsed - lastStatusAt >= STATUS_INTERVAL_MS) {
                    lastStatusAt = nowElapsed;
                    sendStatus(true, "monitoring");
                }
            }
        } catch (SecurityException denied) {
            failMonitoring("Microphone permission lost", "permission_lost");
        } catch (IllegalStateException audioFailure) {
            failMonitoring("Audio monitoring stopped", "audio_error");
        }
    }

    private void deliverAlertWithRetry(double rms) {
        String alertId = UUID.randomUUID().toString();
        Thread retry = new Thread(() -> {
            for (int attempt = 0; attempt < MAX_SEND_ATTEMPTS; attempt++) {
                if (!running || isAcknowledged(alertId)) return;
                sendToConnectedNodes(
                        Protocol.NOISE_PATH,
                        Protocol.encodeAlert(alertId, Protocol.ALERT_TYPE_NOISE, rms)
                );
                if (attempt < MAX_SEND_ATTEMPTS - 1) {
                    try {
                        Thread.sleep(RETRY_INTERVAL_MS);
                    } catch (InterruptedException interrupted) {
                        Thread.currentThread().interrupt();
                        return;
                    }
                }
            }
            if (!isAcknowledged(alertId)) {
                updateNotification("Alert not confirmed — check phone");
            }
        }, "baby-alert-retry");
        retry.start();
    }

    private boolean isAcknowledged(String alertId) {
        return alertId.equals(
                getSharedPreferences(WatchMessageListenerService.PREFS, MODE_PRIVATE)
                        .getString(WatchMessageListenerService.KEY_LAST_ACK, null)
        );
    }

    private void sendStatus(boolean monitoring, String state) {
        int battery = batteryPercent();
        Wearable.getNodeClient(this).getConnectedNodes().addOnSuccessListener(nodes -> {
            boolean connected = !nodes.isEmpty();
            if (!connected && !warnedDisconnected) {
                warnedDisconnected = true;
                updateNotification("Phone disconnected");
            } else if (connected && warnedDisconnected) {
                warnedDisconnected = false;
                updateNotification(monitoring ? "Monitoring • reconnected" : "Stopped");
            }

            if (battery >= 0 && battery <= LOW_BATTERY_PERCENT && !warnedLowBattery) {
                warnedLowBattery = true;
                updateNotification("Low battery: " + battery + "%");
            } else if (battery >= LOW_BATTERY_RESET_PERCENT) {
                warnedLowBattery = false;
            }

            String value = "battery=" + battery
                    + ",connected=" + connected
                    + ",monitoring=" + monitoring
                    + ",state=" + state
                    + ",time=" + System.currentTimeMillis();
            byte[] payload = value.getBytes(StandardCharsets.UTF_8);
            for (com.google.android.gms.wearable.Node node : nodes) {
                Wearable.getMessageClient(this).sendMessage(node.getId(), Protocol.STATUS_PATH, payload);
            }
        });
    }

    private void sendToConnectedNodes(String path, byte[] payload) {
        Wearable.getNodeClient(this).getConnectedNodes().addOnSuccessListener(nodes -> {
            for (com.google.android.gms.wearable.Node node : nodes) {
                Wearable.getMessageClient(this).sendMessage(node.getId(), path, payload);
            }
        });
    }

    private int batteryPercent() {
        Intent battery = registerReceiver(null, new IntentFilter(Intent.ACTION_BATTERY_CHANGED));
        if (battery == null) return -1;
        int level = battery.getIntExtra(BatteryManager.EXTRA_LEVEL, -1);
        int scale = battery.getIntExtra(BatteryManager.EXTRA_SCALE, 100);
        return level >= 0 && scale > 0 ? level * 100 / scale : -1;
    }

    private double calculateNormalizedRms(short[] samples, int count) {
        double sumSquares = 0.0;
        for (int index = 0; index < count; index++) {
            double normalized = samples[index] / 32768.0;
            sumSquares += normalized * normalized;
        }
        return Math.sqrt(sumSquares / count);
    }

    private Notification notification(String text) {
        PendingIntent openApp = PendingIntent.getActivity(
                this,
                0,
                new Intent(this, MainActivity.class),
                PendingIntent.FLAG_UPDATE_CURRENT | PendingIntent.FLAG_IMMUTABLE
        );
        return new Notification.Builder(this, CHANNEL_ID)
                .setSmallIcon(android.R.drawable.ic_btn_speak_now)
                .setContentTitle("Baby monitor active")
                .setContentText(text)
                .setOngoing(true)
                .setOnlyAlertOnce(true)
                .setSilent(true)
                .setContentIntent(openApp)
                .build();
    }

    private void updateNotification(String text) {
        NotificationManager manager = getSystemService(NotificationManager.class);
        if (manager != null) manager.notify(NOTIFICATION_ID, notification(text));
    }

    private void createSilentChannel() {
        NotificationChannel channel = new NotificationChannel(
                CHANNEL_ID,
                "Baby monitor monitoring",
                NotificationManager.IMPORTANCE_LOW
        );
        channel.setSound(null, null);
        channel.enableVibration(false);
        channel.setVibrationPattern(null);
        NotificationManager manager = getSystemService(NotificationManager.class);
        if (manager != null) manager.createNotificationChannel(channel);
    }

    private void stopRecorderSafely() {
        if (recorder == null) return;
        try {
            if (recorder.getRecordingState() == AudioRecord.RECORDSTATE_RECORDING) {
                recorder.stop();
            }
        } catch (IllegalStateException ignored) {
            // Recorder was already stopped.
        }
    }

    private void failMonitoring(String visibleStatus, String state) {
        updateNotification(visibleStatus);
        sendStatus(false, state);
        stopSelf();
    }
}

File: watch/src/main/java/com/example/wearbabymonitor/NoiseMonitorService.kt
package com.example.wearbabymonitor

import android.app.NotificationChannel
import android.app.NotificationManager
import android.app.PendingIntent
import android.app.Service
import android.content.Intent
import android.content.IntentFilter
import android.media.AudioFormat
import android.media.AudioRecord
import android.media.MediaRecorder
import android.os.BatteryManager
import android.os.IBinder
import android.os.SystemClock
import androidx.core.app.NotificationCompat
import com.google.android.gms.wearable.Wearable
import java.util.Locale
import java.util.UUID
import kotlin.math.max
import kotlin.math.sqrt

class NoiseMonitorService : Service() {
    @Volatile private var running = false
    private var recorder: AudioRecord? = null
    private var worker: Thread? = null
    private var lastAlertAt = 0L
    private var calibratedThreshold = DEFAULT_THRESHOLD
    private var lastStatusAt = 0L
    private var warnedDisconnected = false
    private var warnedLowBattery = false

    override fun onCreate() {
        super.onCreate()
        createChannel()
    }

    override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int {
        startForeground(NOTIFICATION_ID, notification("Calibrating room noise…"))
        if (!running) startRecordingLoop()
        return START_NOT_STICKY
    }

    override fun onDestroy() {
        val wasRunning = running
        running = false
        recorder?.stopSafely()
        recorder?.release()
        recorder = null
        worker?.interrupt()
        worker = null
        if (wasRunning) {
            sendStatus(monitoring = false, state = "stopped")
        }
        super.onDestroy()
    }

    override fun onBind(intent: Intent?): IBinder? = null

    private fun startRecordingLoop() {
        val sampleRate = 16_000
        val minBuffer = AudioRecord.getMinBufferSize(
            sampleRate,
            AudioFormat.CHANNEL_IN_MONO,
            AudioFormat.ENCODING_PCM_16BIT
        )
        if (minBuffer <= 0) {
            updateNotification("Microphone unavailable")
            sendStatus(monitoring = false, state = "microphone_unavailable")
            stopSelf()
            return
        }

        val bufferSize = max(minBuffer, 2048)
        val audioRecord = AudioRecord(
            MediaRecorder.AudioSource.MIC,
            sampleRate,
            AudioFormat.CHANNEL_IN_MONO,
            AudioFormat.ENCODING_PCM_16BIT,
            bufferSize * 2
        )
        if (audioRecord.state != AudioRecord.STATE_INITIALIZED) {
            audioRecord.release()
            updateNotification("Microphone unavailable")
            sendStatus(monitoring = false, state = "microphone_unavailable")
            stopSelf()
            return
        }

        recorder = audioRecord
        running = true
        worker = Thread {
            val buffer = ShortArray(bufferSize)
            val calibrationSamples = mutableListOf<Double>()
            val calibrationEnd = SystemClock.elapsedRealtime() + CALIBRATION_MS
            var loudWindows = 0

            try {
                audioRecord.startRecording()
                sendStatus(monitoring = true, state = "calibrating")
                lastStatusAt = SystemClock.elapsedRealtime()

                while (running && !Thread.currentThread().isInterrupted) {
                    val count = audioRecord.read(buffer, 0, buffer.size)
                    if (count <= 0) continue
                    val rms = calculateNormalizedRms(buffer, count)
                    val nowElapsed = SystemClock.elapsedRealtime()

                    if (nowElapsed < calibrationEnd) {
                        calibrationSamples += rms
                        continue
                    } else if (calibrationSamples.isNotEmpty()) {
                        val sorted = calibrationSamples.sorted()
                        val baseline = sorted[(sorted.size * 0.8).toInt().coerceAtMost(sorted.lastIndex)]
                        calibratedThreshold = max(MIN_THRESHOLD, baseline * BASELINE_MULTIPLIER)
                        calibrationSamples.clear()
                        updateNotification(
                            "Monitoring • threshold ${String.format(Locale.US, "%.3f", calibratedThreshold)}"
                        )
                        sendStatus(monitoring = true, state = "monitoring")
                    }

                    loudWindows = if (rms >= calibratedThreshold) loudWindows + 1 else 0
                    if (loudWindows >= REQUIRED_LOUD_WINDOWS) {
                        val now = System.currentTimeMillis()
                        if (now - lastAlertAt >= ALERT_COOLDOWN_MS) {
                            lastAlertAt = now
                            deliverAlertWithRetry(rms)
                        }
                        loudWindows = 0
                    }

                    if (nowElapsed - lastStatusAt >= STATUS_INTERVAL_MS) {
                        lastStatusAt = nowElapsed
                        sendStatus(monitoring = true, state = "monitoring")
                    }
                }
            } catch (_: SecurityException) {
                updateNotification("Microphone permission lost")
                sendStatus(monitoring = false, state = "permission_lost")
                stopSelf()
            } catch (_: IllegalStateException) {
                updateNotification("Audio monitoring stopped")
                sendStatus(monitoring = false, state = "audio_error")
                stopSelf()
            }
        }.apply {
            name = "baby-noise-monitor"
            start()
        }
    }

    private fun deliverAlertWithRetry(rms: Double) {
        val alertId = UUID.randomUUID().toString()
        Thread {
            repeat(MAX_SEND_ATTEMPTS) { attempt ->
                if (!running || isAcknowledged(alertId)) return@Thread
                sendToConnectedNodes(
                    Protocol.NOISE_PATH,
                    Protocol.encodeAlert(
                        id = alertId,
                        type = Protocol.ALERT_TYPE_NOISE,
                        rms = rms
                    )
                )
                if (attempt < MAX_SEND_ATTEMPTS - 1) Thread.sleep(RETRY_INTERVAL_MS)
            }
            if (!isAcknowledged(alertId)) {
                // Visual-only status. The watch channel has sound and vibration disabled.
                updateNotification("Alert not confirmed — check phone")
            }
        }.apply {
            name = "baby-alert-retry"
            start()
        }
    }

    private fun isAcknowledged(alertId: String): Boolean =
        getSharedPreferences(WatchMessageListenerService.PREFS, MODE_PRIVATE)
            .getString(WatchMessageListenerService.KEY_LAST_ACK, null) == alertId

    private fun sendStatus(monitoring: Boolean, state: String) {
        val battery = batteryPercent()
        Wearable.getNodeClient(this).connectedNodes
            .addOnSuccessListener { nodes ->
                val connected = nodes.isNotEmpty()
                if (!connected && !warnedDisconnected) {
                    warnedDisconnected = true
                    updateNotification("Phone disconnected")
                } else if (connected && warnedDisconnected) {
                    warnedDisconnected = false
                    updateNotification(if (monitoring) "Monitoring • reconnected" else "Stopped")
                }

                if (battery in 0..LOW_BATTERY_PERCENT && !warnedLowBattery) {
                    warnedLowBattery = true
                    updateNotification("Low battery: $battery%")
                } else if (battery >= LOW_BATTERY_RESET_PERCENT) {
                    warnedLowBattery = false
                }

                val payload = buildString {
                    append("battery=").append(battery)
                    append(",connected=").append(connected)
                    append(",monitoring=").append(monitoring)
                    append(",state=").append(state)
                    append(",time=").append(System.currentTimeMillis())
                }.encodeToByteArray()
                nodes.forEach { node ->
                    Wearable.getMessageClient(this)
                        .sendMessage(node.id, Protocol.STATUS_PATH, payload)
                }
            }
    }

    private fun sendToConnectedNodes(path: String, payload: ByteArray) {
        Wearable.getNodeClient(this).connectedNodes.addOnSuccessListener { nodes ->
            nodes.forEach { node ->
                Wearable.getMessageClient(this).sendMessage(node.id, path, payload)
            }
        }
    }

    private fun batteryPercent(): Int {
        val intent = registerReceiver(null, IntentFilter(Intent.ACTION_BATTERY_CHANGED)) ?: return -1
        val level = intent.getIntExtra(BatteryManager.EXTRA_LEVEL, -1)
        val scale = intent.getIntExtra(BatteryManager.EXTRA_SCALE, 100)
        return if (level >= 0 && scale > 0) (level * 100 / scale) else -1
    }

    private fun calculateNormalizedRms(samples: ShortArray, count: Int): Double {
        var sumSquares = 0.0
        for (i in 0 until count) {
            val normalized = samples[i] / 32768.0
            sumSquares += normalized * normalized
        }
        return sqrt(sumSquares / count)
    }

    private fun notification(text: String) = NotificationCompat.Builder(this, CHANNEL_ID)
        .setSmallIcon(android.R.drawable.ic_btn_speak_now)
        .setContentTitle("Baby monitor active")
        .setContentText(text)
        .setOngoing(true)
        .setOnlyAlertOnce(true)
        .setSilent(true)
        .setContentIntent(
            PendingIntent.getActivity(
                this,
                0,
                Intent(this, MainActivity::class.java),
                PendingIntent.FLAG_UPDATE_CURRENT or PendingIntent.FLAG_IMMUTABLE
            )
        )
        .build()

    private fun updateNotification(text: String) {
        getSystemService(NotificationManager::class.java)
            .notify(NOTIFICATION_ID, notification(text))
    }

    private fun createChannel() {
        val channel = NotificationChannel(
            CHANNEL_ID,
            "Baby monitor monitoring",
            NotificationManager.IMPORTANCE_LOW
        ).apply {
            setSound(null, null)
            enableVibration(false)
            vibrationPattern = null
        }
        getSystemService(NotificationManager::class.java)
            .createNotificationChannel(channel)
    }

    private fun AudioRecord.stopSafely() {
        try {
            if (recordingState == AudioRecord.RECORDSTATE_RECORDING) stop()
        } catch (_: IllegalStateException) {
            // Already stopped.
        }
    }

    companion object {
        private const val CHANNEL_ID = "baby_monitor_running_v2"
        private const val NOTIFICATION_ID = 1
        private const val DEFAULT_THRESHOLD = 0.075
        private const val MIN_THRESHOLD = 0.018
        private const val BASELINE_MULTIPLIER = 3.0
        private const val CALIBRATION_MS = 8_000L
        private const val REQUIRED_LOUD_WINDOWS = 3
        private const val ALERT_COOLDOWN_MS = 10_000L
        private const val STATUS_INTERVAL_MS = 10_000L
        private const val MAX_SEND_ATTEMPTS = 6
        private const val RETRY_INTERVAL_MS = 1_500L
        private const val LOW_BATTERY_PERCENT = 20
        private const val LOW_BATTERY_RESET_PERCENT = 25
    }
}
